### DISCUSSION TITLE
Trigger a function between retries

### DISCUSSION THREAD

**Discussion by Jrmromao** [resilience4j/resilience4j#1983] [Q&A]
Hi all,

I have recently started integrating the Resilience4j library into one of our projects. However, I have encountered an issue and I'm not sure if this library is the best choice for my requirements.

I am utilizing the retry functionality for an HTTP call to an external API, where I send an authentication token for authorization. In the event of a failed HTTP call, the system will retry sending the request three times. However, the token remains unchanged during these retries. Ideally, I would like to update the token even if it is still valid.

Could you please advise on the best approach to achieve this functionality?

Thank you in advance.

**Comment by RobWin** [comment]
Hi,
our implementation won't take care of this, since it's protocol agnostic.
Your method/function, which is decorated by the retry functionality, must internally make sure that it retrieves a new/valid token before calling the external API.
It shouldn't matter if the function is called because a prev. call failed or not.

**Reply by Jrmromao** [reply]
Thanks for the quick reply. I was able to accomplish what I needed using the RetryRegistry and getEventPublisher.

Now that I'm writing the unit tests, I've encountered a new problem. I'm initializing the RetryRegistry in the class constructor, and when I run the tests, I receive the following error:

"You haven't provided the instance at field declaration, so I tried to construct the instance. However, the constructor or the initialization block threw an exception: Cannot invoke 'io.github.resilience4j.retry.RetryRegistry.retry(String)' because 'retryRegistry' is null."

I thought I could initialize the RetryRegistry by using a mock, but it doesn't seem to work.

Do you have any ideas on what I might be doing wrong?

**Reply by RobWin** [reply]
I need more context, maybe an example of your test.
Are you using Spring Boot? Our Sprin Boot Starter?
